test/file-transfer-handler.test.js

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const { createFile  } = require('./test-object-helper');
+const { importFile, getTestOptions } = require('./testutils');
+const PartUploadResult = importFile('part-upload-result');
+const HttpClient = importFile('http/http-client');
+const FileTransferHandler = importFile('file-transfer-handler');
+const DirectBinaryUploadOptions = importFile('direct-binary-upload-options');
+const UploadResult = importFile('upload-result');
+const { setTimeoutPromise } = importFile('utils');
+
+const FILE_NAME = 'testfile.jpg';
+const FILE_PATH = `/content/dam/${FILE_NAME}`;
+
+describe('FileTransferHandler tests', function() {
+    let handler = false;
+    let allFileParts = [];
+    let initResponseFileInfo = false;
+    let options = false;
+    let uploadOptions = false;
+    let httpClient = false;
+    let transferStarted = [];
+    let transferProgress = [];
+    let transferCancelled = [];
+    let transferError = [];
+    let transferSuccess = [];
+
+    beforeEach(function() {
+        options = getTestOptions();
+        uploadOptions = new DirectBinaryUploadOptions();
+        httpClient = new HttpClient(options, uploadOptions);
+        handler = new FileTransferHandler(options, uploadOptions, httpClient);
+        const file = createFile(options, uploadOptions, `/${FILE_NAME}`, 'testing', 2);
+        allFileParts = file.getParts();
+        initResponseFileInfo = allFileParts[0];
+        transferStarted = [];
+        transferProgress = [];
+        transferCancelled = [];
+        transferError = [];
+        transferSuccess = [];
+    
+        handler._doFileTransferStarted = async(fileUploadResult, fileInfo) => {
```

Comment:
From the class definition it's not clear those functions should be overriden. Either add a log that it's using an empty function, or throw an error in the parent class.

Reply:
Yeah, it would sure be nice if JS could do abstract classes.

The method docs indicate that the default implementation doesn't do anything, but I went ahead and added a `warn` log message if the empty implementation is called.